src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs

```diff
@@ -315,51 +320,23 @@ private static bool SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol,
-                {
-                    // SymbolEquivalenceComparer performs Location equality checks for Locals, Labels and RangeVariables.
-                    // As we are comparing symbols from different semantic models, locations will differ.
-                    // Hence perform minimal checks for these symbol kinds.
-                    case SymbolKind.Local:
-                        return newSymbol.Kind == SymbolKind.Local &&
-                            newSymbol.IsImplicitlyDeclared == symbol.IsImplicitlyDeclared &&
-                            string.Equals(symbol.Name, newSymbol.Name, StringComparison.Ordinal) &&
-                            ((ILocalSymbol)newSymbol).Type.Equals(((ILocalSymbol)symbol).Type);
-
-                    case SymbolKind.Label:
-                    case SymbolKind.RangeVariable:
-                        return newSymbol.Kind == symbol.Kind && string.Equals(newSymbol.Name, symbol.Name, StringComparison.Ordinal);
-
-                    case SymbolKind.Parameter:
-                        if (newSymbol.Kind == SymbolKind.Parameter && symbol.ContainingSymbol.IsAnonymousFunction())
-                        {
-                            var param = (IParameterSymbol)symbol;
-                            var newParam = (IParameterSymbol)newSymbol;
-                            return param.IsRefOrOut() == newParam.IsRefOrOut() &&
-                                param.Name == newParam.Name &&
-                                SymbolEquivalenceComparer.Instance.Equals(param.Type, newParam.Type) &&
-                                newSymbol.ContainingSymbol.IsAnonymousFunction();
-                        }
-
-                        goto default;
-
-                    default:
-                        return SymbolEquivalenceComparer.Instance.Equals(symbol, newSymbol);
-                }
+                return CompareAcrossSemanticModels(symbol, newSymbol);
             }
 
             if (symbol.Equals(newSymbol, SymbolEqualityComparer.IncludeNullability))
             {
                 return true;
             }
 
-            // Handle equivalence of special built-in comparison operators between enum types and 
-            // it's underlying enum type.
-            if (symbol.Kind == SymbolKind.Method && newSymbol.Kind == SymbolKind.Method)
+            if (symbol is IMethodSymbol methodSymbol && newSymbol is IMethodSymbol newMethodSymbol)
             {
-                var methodSymbol = (IMethodSymbol)symbol;
-                var newMethodSymbol = (IMethodSymbol)newSymbol;
+                // If we have local functions, we can't use normal symbol equality for them (since that checks locations).
+                // Have to defer to SymbolEquivalence instead.
+                if (methodSymbol.MethodKind == MethodKind.LocalFunction && newMethodSymbol.MethodKind == MethodKind.LocalFunction)
+                    return CompareAcrossSemanticModels(methodSymbol, newMethodSymbol);
```

Comment:
❔ Can you clarify which of the bugs this change fixed? Can be a reply to this comment and not in code if there are already tests covering regressions.

Reply:
mentioned above.